checkout-core/src/main/java/com/adyen/checkout/core/util/FileDownloader.kt

```diff
@@ -0,0 +1,117 @@
+import java.net.URI
+import java.net.URL
+
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+@Suppress("LongParameterList")
+class FileDownloader(private val context: Context) {
+
+    @RequiresPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+    suspend fun download(
+        stringUrl: String,
+        fileName: String,
+        filePath: String,
+        mimeType: String? = null
+    ): Result<Unit> {
+        val url = stringUrl.toURL() ?: return Result.failure(CheckoutException("Malformed URL"))
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+            downloadAndSaveApi29AndAbove(context, url, fileName, filePath, mimeType ?: "*/*")
+        } else {
+            downloadAndSaveApi28AndBelow(context, url, fileName, filePath)
+        }
+    }
+
+    @RequiresPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+    private suspend fun downloadAndSaveApi28AndBelow(
+        context: Context,
+        url: URL,
+        fileName: String,
+        filePath: String,
+    ): Result<Unit> = withContext(Dispatchers.IO) {
+        val imageFile = File(context.getExternalFilesDir(filePath), fileName)
+        url.openStream().use { input ->
+            FileOutputStream(imageFile).use { output ->
+                input.copyTo(output)
+
+                val insertImageResult = MediaStore.Images.Media.insertImage(
+                    context.contentResolver,
+                    imageFile.absolutePath,
+                    fileName,
+                    null
+                )
+
+                return@withContext if (insertImageResult == null) {
+                    Result.failure(CheckoutException("couldn't insert image to gallery"))
+                } else {
+                    Result.success(Unit)
+                }
+            }
+        }
+    }
+
+    @RequiresApi(Build.VERSION_CODES.Q)
+    private suspend fun downloadAndSaveApi29AndAbove(
+        context: Context,
+        url: URL,
+        fileName: String,
+        filePath: String,
+        mimeType: String
+    ): Result<Unit> = withContext(Dispatchers.IO) {
+        val contentValues = ContentValues().apply {
+            put(MediaStore.MediaColumns.DISPLAY_NAME, fileName)
+            put(MediaStore.MediaColumns.MIME_TYPE, mimeType)
+            put(MediaStore.MediaColumns.RELATIVE_PATH, filePath)
+        }
+        val resolver = context.contentResolver
+        val uri = resolver.insert(MediaStore.Downloads.EXTERNAL_CONTENT_URI, contentValues)
+            ?: return@withContext Result.failure(CheckoutException("URI is null"))
+
+        url.openStream().use { input ->
+            resolver.openOutputStream(uri).use { output ->
+                if (output == null) return@withContext Result.failure(CheckoutException("out is null"))
+                input.copyTo(output, DEFAULT_BUFFER_SIZE)
+                return@withContext Result.success(Unit)
+            }
+        }
+    }
+
+    private fun String.toURL(): URL? {
```

Comment:
Did you try `Uri.encode`?

Reply:
No, will  check it.

Reply:
Didn't work I got `OutOfMemoryError`

Reply:
I got this error after doing the encoding
```
java.net.MalformedURLException: Protocol not found: https%3A%2F%2Fcheckoutshopper-test.adyen.com%2Fcheckoutshopper%2Fbarcode.shtml%3FbarcodeType%3DqrCode%26fileType%3Dpng%26data%3D00020126580009SG.PAYNOW010120213201210678MADY03010041420230120235031520400005303702540513.375802SG5925ADYEN%20SINGAPORE%20PTE.%20LTD.6009Singapore62200116PWRHRMBJVXSKGK8263043D50
```

Reply:
I got this error after trying it `java.net.MalformedURLException: Protocol not found`